src/main/java/com/shapesecurity/functional/data/ConcatList.java

```diff
@@ -408,6 +509,79 @@ public T next() {
         }
     }
 
+    private final static class ConcatListSplitIterator<T> implements Spliterator<T> {
+        @NotNull
+        private final Deque<ConcatList<T>> stack;
+        private int size;
+
+        private ConcatListSplitIterator(@NotNull Deque<ConcatList<T>> stack, int size) {
+            this.stack = stack;
+            this.size = size;
+        }
+
+
+        private ConcatListSplitIterator(@NotNull ConcatList<T> list) {
+            // Manually expanded recursion
+            stack = new ArrayDeque<>(list.length);
+            stack.add(list);
+            this.size = list.length;
+        }
+
+        @Override
+        public boolean tryAdvance(Consumer<? super T> action) {
+            while (!stack.isEmpty()) {
+                ConcatList<T> curr = stack.pop();
+                if (curr instanceof Fork) {
+                    Fork<T> fork = (Fork<T>) curr;
+                    stack.push(fork.right);
+                    stack.push(fork.left);
+                } else if (curr instanceof Leaf) {
+                    this.size--;
+                    action.accept(((Leaf<T>) curr).data);
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public Spliterator<T> trySplit() {
+            Deque<ConcatList<T>> newDequeue = new ArrayDeque<>();
+            newDequeue.addAll(stack);
+            return new ConcatListSplitIterator<>(newDequeue, size);
+        }
+
+        @Override
+        public long estimateSize() {
+            return size;
+        }
+
+        @Override
+        public int characteristics() {
```

Comment:
This comes from the new API of `Iterable`. This can be used to speed up `Stream`s.